src/C4EnergyBars.cpp

```diff
@@ -0,0 +1,586 @@
+	std::string file;
+	try {
+		auto gfx = gfxs.at(key);
+		amount = gfx.amount;
+		scale = gfx.scale;
+		file = gfx.file;
+	} catch (const std::out_of_range &) {
+		LogF("DefineEnergyBars: Missing key '%s' in graphics definition", key.c_str());
+		return nullptr;
+	}
+
+	// TODO FIXME why wasn't this already called before we get here?
+	Game.GraphicsResource.RegisterMainGroups();
+
+	auto facet = std::shared_ptr<C4FacetExID>(new C4FacetExID(), [=](C4FacetExID* facet){graphics.erase(key);});
+	bool success = Game.GraphicsResource.LoadFile(*facet, file.c_str(), Game.GraphicsResource.Files);
+	if(!success) {LogF("could not load custom energy bar graphic '%s'", file.c_str()); return nullptr;}
+
+	int32_t bar_wdt = facet->Surface->Wdt / (amount * 2);
+	int32_t bar_hgt = facet->Surface->Hgt / 3;
+	bar_wdt = (bar_wdt*100) / scale;
+	bar_hgt = (bar_hgt*100) / scale;
+	facet->Set(facet->Surface, 0, 0, bar_wdt, bar_hgt);
+
+	graphics.emplace(key, std::weak_ptr<C4FacetExID>(facet));
+	return facet;
+}
+
+std::shared_ptr<C4EnergyBarsDef> C4EnergyBarsUniquifier::UniqueifyDefinition(C4EnergyBarsDef *definition)
+{
+	// TODO somehow change *definition to unique_ptr?
+	// currently definition always gets owned by a shared_ptr, that either ends up being the canonical shared_ptr
+	// or goes out of scope deleting the definition
+
+	// the weak ptr remembers the custom deleter
+	auto shared = std::shared_ptr<C4EnergyBarsDef>(definition, [=](C4EnergyBarsDef* def){definitions.erase(*def);});
+	auto it_success = definitions.emplace(*definition, std::weak_ptr<C4EnergyBarsDef>(shared));
+	if (!it_success.second) {
+		// definition already existed, we have to override shared so that it is linked to
+		// the same control block that was used to create the first shared ptr
+		shared = it_success.first->second.lock();
+	}
+	return shared;
+}
+
+std::shared_ptr<C4EnergyBars> C4EnergyBarsUniquifier::DefineEnergyBars(C4ValueHash* graphics, C4ValueArray *definition)
+{
+	if (!graphics) {
+		LogF("DefineEnergyBars Parameter 1 map expected but got: %s", C4Value(graphics).GetDataString().getData());
+		return nullptr;
+	}
+
+	if (!definition) {
+		LogF("DefineEnergyBars Parameter 2 array expected but got: %s", C4Value(definition).GetDataString().getData());
+		return nullptr;
+	}
+
+	int32_t value_index = 0;
+	C4EnergyBarsDef::Gfxs gfx;
+	C4EnergyBarsDef::Bars bars;
+	C4EnergyBarsDef::Names names;
+	bool success = ProcessGraphics(*graphics, gfx) && ProcessGroup(value_index, gfx, *definition, bars, true) && C4EnergyBarsDef::PopulateNamesFromValues(bars, names);
+	if(success) {
+		auto def = UniqueifyDefinition(new C4EnergyBarsDef(gfx, bars, names));
+		return Instantiate(def);
+	}
+	return nullptr;
+}
+
+bool C4EnergyBarsUniquifier::ProcessGraphics(C4ValueHash &map, C4EnergyBarsDef::Gfxs &gfx)
+{
+	auto amount = C4VString("amount");
+	auto scale  = C4VString("scale");
+
+	C4ValueHash::Iterator end = map.end();
+	for (C4ValueHash::Iterator it = map.begin(); it != end; ++it) {
```

Comment:
why not `for(const auto &[key, val] : map)`?

Reply:
even without const: `for(auto &[key, val] : map)` I run into the problem that key ends up being const and compiler then refuses call to `key.getStr`, it works if I use `key._getStr`, but that seems wrong because I can't assume that I actually get a string from the script. I am unsure whether there should be a const variant of `C4Value::getStr` or key should end up being non const, possibly by changing how `C4ValueHash` works.

Reply:
Hmm, so without changing semantics without a reference maybe?

But when thinking about it, it seems like you expect specifically strings as key type. In that case it seems more reasonable to check the type and error out for wrong types and then you can also use the _getStr(). Or is there some intended conversion involved that I missed?

Either way, the current way actually looks dangerous to me, in case that getStr() returns nullptr.
Same for getMap().

Reply:
Currently I explicitly copy the key to get it to work, maybe that can be expressed more succinctly somehow.

Reply:
my idea was basically the following:
```
for (const auto &[key, val] : map)
{
    if (key.GetType() != C4V_String) type_error;
    const auto keyStr = key._getStr();

    // same thing for map
}
```

Reply:
I used GetType() first, but seems type can be Any so I used ConvertTo so that an any that happens to be a string works too.

Reply:
The whole any type thing is so misleading. In practice only nil has type any.

Reply:
So GetType works now, must have done something wrong before.
But I think the code might be still wrong if someone uses a reference as a key?
Because GetType returns the type of the referenced value but _getStr() would then be called on the reference not the referenced? Should I call key.GetRefVal()._getStr()?

Reply:
Good point. `key.GetRefVal()._getStr()` seems like the right choice.
Although, currently you can not store a reference in a map, it still is the better way, I think.